File: backend/app/api/v1/endpoints/categories.py
"""
Category endpoints
"""
from fastapi import APIRouter, Depends, HTTPException, status, Query
from typing import List, Optional
from app.api.v1.schemas.category import (
    CategoryCreate, CategoryUpdate, CategoryResponse
)
from app.middleware.auth import get_current_user, get_current_admin
from app.db.database import get_database
from bson import ObjectId
from datetime import datetime

router = APIRouter()


@router.get("/", response_model=List[CategoryResponse])
async def get_categories(
    organization: Optional[str] = Query(None),
    current_user: dict = Depends(get_current_user)
):
    """Get all categories"""
    db = await get_database()
    query = {"is_active": True}
    
    if organization:
        query["organization"] = ObjectId(organization)
    elif current_user.get("organization"):
        query["organization"] = ObjectId(current_user["organization"])
    
    cursor = db.categories.find(query)
    categories = await cursor.to_list(length=100)
    
    result = []
    for cat in categories:
        cat["id"] = str(cat["_id"])
        if cat.get("organization"):
            cat["organization"] = str(cat["organization"])
        del cat["_id"]
        result.append(cat)
    
    return result


@router.get("/all", response_model=List[CategoryResponse])
async def get_all_categories(
    organization: Optional[str] = Query(None),
    current_user: dict = Depends(get_current_admin)
):
    """Get all categories including inactive (admin only)"""
    db = await get_database()
    query = {}
    
    if organization:
        query["organization"] = ObjectId(organization)
    
    cursor = db.categories.find(query)
    categories = await cursor.to_list(length=100)
    
    result = []
    for cat in categories:
        cat["id"] = str(cat["_id"])
        if cat.get("organization"):
            cat["organization"] = str(cat["organization"])
        del cat["_id"]
        result.append(cat)
    
    return result


@router.get("/{category_id}", response_model=CategoryResponse)
async def get_category(category_id: str, current_user: dict = Depends(get_current_user)):
    """Get category by ID"""
    db = await get_database()
    
    try:
        category = await db.categories.find_one({"_id": ObjectId(category_id)})
    except:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Category not found"
        )
    
    if not category:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Category not found"
        )
    
    category["id"] = str(category["_id"])
    if category.get("organization"):
        category["organization"] = str(category["organization"])
    del category["_id"]
    
    return category


@router.post("/", response_model=CategoryResponse)
async def create_category(
    category_data: CategoryCreate,
    current_user: dict = Depends(get_current_admin)
):
    """Create new category (admin only)"""
    db = await get_database()
    
    category_doc = {
        "name": category_data.name,
        "description": category_data.description,
        "is_active": True,
        "created_at": datetime.utcnow()
    }
    
    if category_data.organization:
        category_doc["organization"] = ObjectId(category_data.organization)
    elif current_user.get("organization"):
        category_doc["organization"] = ObjectId(current_user["organization"])
    
    result = await db.categories.insert_one(category_doc)
    category_doc["id"] = str(result.inserted_id)
    if category_doc.get("organization"):
        category_doc["organization"] = str(category_doc["organization"])
    del category_doc["_id"]
    
    return category_doc


@router.put("/{category_id}", response_model=CategoryResponse)
async def update_category(
    category_id: str,
    category_data: CategoryUpdate,
    current_user: dict = Depends(get_current_admin)
):
    """Update category (admin only)"""
    db = await get_database()
    
    update_doc = {}
    if category_data.name:
        update_doc["name"] = category_data.name
    if category_data.description is not None:
        update_doc["description"] = category_data.description
    if category_data.is_active is not None:
        update_doc["is_active"] = category_data.is_active
    
    await db.categories.update_one(
        {"_id": ObjectId(category_id)},
        {"$set": update_doc}
    )
    
    category = await db.categories.find_one({"_id": ObjectId(category_id)})
    category["id"] = str(category["_id"])
    if category.get("organization"):
        category["organization"] = str(category["organization"])
    del category["_id"]
    
    return category


@router.delete("/{category_id}")
async def delete_category(category_id: str, current_user: dict = Depends(get_current_admin)):
    """Delete category (admin only)"""
    db = await get_database()
    
    await db.categories.update_one(
        {"_id": ObjectId(category_id)},
        {"$set": {"is_active": False}}
    )
    
    return {"message": "Category deleted successfully"}

File: server/routes/categories.js
import express from 'express'
import Category from '../models/Category.js'
import Ticket from '../models/Ticket.js'
import { protect, admin } from '../middleware/auth.js'

const router = express.Router()

// @route   GET /api/categories
// @desc    Get all categories (global + organization-specific)
// @access  Private
router.get('/', protect, async (req, res) => {
  try {
    const query = {
      $or: [
        { organization: null }, // Global categories
        { organization: req.user.organization }, // Organization-specific
      ],
      status: 'active',
    }
    
    const categories = await Category.find(query).sort({ name: 1 })
    res.json(categories)
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

// @route   GET /api/categories/all
// @desc    Get all categories including inactive (admin only)
// @access  Private/Admin
router.get('/all', protect, admin, async (req, res) => {
  try {
    const { organization } = req.query
    const query = {}
    
    if (organization) {
      query.$or = [
        { organization: null },
        { organization },
      ]
    } else {
      query.$or = [
        { organization: null },
        { organization: req.user.organization },
      ]
    }
    
    const categories = await Category.find(query).sort({ name: 1 })
    res.json(categories)
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

// @route   GET /api/categories/:id
// @desc    Get single category
// @access  Private
router.get('/:id', protect, async (req, res) => {
  try {
    const category = await Category.findById(req.params.id)
    if (!category) {
      return res.status(404).json({ message: 'Category not found' })
    }
    res.json(category)
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

// @route   POST /api/categories
// @desc    Create new category
// @access  Private/Admin
router.post('/', protect, admin, async (req, res) => {
  try {
    const { name, description, color, organization, status } = req.body

    const categoryExists = await Category.findOne({ name, organization: organization || null })
    if (categoryExists) {
      return res.status(400).json({ message: 'Category with this name already exists' })
    }

    const category = await Category.create({
      name,
      description: description || '',
      color: color || '#00ffff',
      organization: organization || null, // null = global
      status: status || 'active',
    })

    res.status(201).json(category)
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

// @route   PUT /api/categories/:id
// @desc    Update category
// @access  Private/Admin
router.put('/:id', protect, admin, async (req, res) => {
  try {
    const category = await Category.findById(req.params.id)
    if (!category) {
      return res.status(404).json({ message: 'Category not found' })
    }

    const { name, description, color, status } = req.body

    if (name && name !== category.name) {
      const categoryExists = await Category.findOne({ 
        name, 
        organization: category.organization,
        _id: { $ne: category._id }
      })
      if (categoryExists) {
        return res.status(400).json({ message: 'Category with this name already exists' })
      }
      category.name = name
    }

    if (description !== undefined) category.description = description
    if (color) category.color = color
    if (status) category.status = status

    await category.save()
    res.json(category)
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

// @route   DELETE /api/categories/:id
// @desc    Delete category
// @access  Private/Admin
router.delete('/:id', protect, admin, async (req, res) => {
  try {
    const category = await Category.findById(req.params.id)
    if (!category) {
      return res.status(404).json({ message: 'Category not found' })
    }

    // Check if category is used in any tickets
    const ticketCount = await Ticket.countDocuments({ category: category.name })
    if (ticketCount > 0) {
      return res.status(400).json({ 
        message: `Cannot delete category. It is used in ${ticketCount} ticket(s). Please reassign those tickets first.` 
      })
    }

    await category.deleteOne()
    res.json({ message: 'Category deleted successfully' })
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

export default router

